From 5e63859748e7b4331348667ad522c1e86946f297 Mon Sep 17 00:00:00 2001
From: Johnny Mnemonic <jm@machine-hall.org>
Date: Fri, 13 Feb 2026 13:55:34 +0100
Subject: [PATCH 097/101] ia64: introduce arch_zone_limits_init()

Move calculations of zone limits to a dedicated arch_zone_limits_init()
function.

Later MM core will use this function as an architecture specific callback
during nodes and zones initialization and thus there won't be a need to
call free_area_init() from every architecture.

Based on the corresponding patches for:
* [alpha](ba1c86874e25e95de9b253570bb50cc3b5df542e)
* [loongarch](63cadcb731c914c85577512b0688fd62350644a8)
---
 arch/ia64/mm/contig.c    | 19 ++++++++++++-------
 arch/ia64/mm/discontig.c | 15 ++++++++++-----
 2 files changed, 22 insertions(+), 12 deletions(-)

diff --git a/arch/ia64/mm/contig.c b/arch/ia64/mm/contig.c
index 1e9eaa107eb7..67fb7dddbc6d 100644
--- a/arch/ia64/mm/contig.c
+++ b/arch/ia64/mm/contig.c
@@ -186,6 +186,17 @@ static void __init verify_gap_absence(void)
 		      (max_gap >> 20));
 }
 
+void __init arch_zone_limits_init(unsigned long *max_zone_pfns)
+{
+	unsigned long max_dma;
+
+	max_dma = virt_to_phys((void *) MAX_DMA_ADDRESS) >> PAGE_SHIFT;
+	max_zone_pfns[ZONE_DMA32] = max_dma;
+	max_zone_pfns[ZONE_NORMAL] = max_low_pfn;
+
+	verify_gap_absence();
+}
+
 /*
  * Set up the page tables.
  */
@@ -193,16 +204,10 @@ static void __init verify_gap_absence(void)
 void __init
 paging_init (void)
 {
-	unsigned long max_dma;
 	unsigned long max_zone_pfns[MAX_NR_ZONES];
 
 	memset(max_zone_pfns, 0, sizeof(max_zone_pfns));
-	max_dma = virt_to_phys((void *) MAX_DMA_ADDRESS) >> PAGE_SHIFT;
-	max_zone_pfns[ZONE_DMA32] = max_dma;
-	max_zone_pfns[ZONE_NORMAL] = max_low_pfn;
-
-	verify_gap_absence();
-
+	arch_zone_limits_init(max_zone_pfns);
 	free_area_init(max_zone_pfns);
 	zero_page_memmap_ptr = virt_to_page(ia64_imva(empty_zero_page));
 }
diff --git a/arch/ia64/mm/discontig.c b/arch/ia64/mm/discontig.c
index 9860b250c845..b8683f5d8034 100644
--- a/arch/ia64/mm/discontig.c
+++ b/arch/ia64/mm/discontig.c
@@ -585,6 +585,15 @@ void call_pernode_memory(unsigned long start, unsigned long len, void *arg)
 	}
 }
 
+void __init arch_zone_limits_init(unsigned long *max_zone_pfns)
+{
+	unsigned long max_dma;
+
+	max_dma = virt_to_phys((void *) MAX_DMA_ADDRESS) >> PAGE_SHIFT;
+	max_zone_pfns[ZONE_DMA32] = max_dma;
+	max_zone_pfns[ZONE_NORMAL] = max_low_pfn;
+}
+
 /**
  * paging_init - setup page tables
  *
@@ -593,16 +602,12 @@ void call_pernode_memory(unsigned long start, unsigned long len, void *arg)
  */
 void __init paging_init(void)
 {
-	unsigned long max_dma;
 	unsigned long max_zone_pfns[MAX_NR_ZONES];
 
-	max_dma = virt_to_phys((void *) MAX_DMA_ADDRESS) >> PAGE_SHIFT;
-
 	sparse_init();
 
 	memset(max_zone_pfns, 0, sizeof(max_zone_pfns));
-	max_zone_pfns[ZONE_DMA32] = max_dma;
-	max_zone_pfns[ZONE_NORMAL] = max_low_pfn;
+	arch_zone_limits_init(max_zone_pfns);
 	free_area_init(max_zone_pfns);
 
 	zero_page_memmap_ptr = virt_to_page(ia64_imva(empty_zero_page));
-- 
2.25.1

